boa_engine/src/optimizer/constant_folding_optimizer.rs

```diff
@@ -0,0 +1,152 @@
+use boa_ast::{
+    expression::{
+        literal::Literal,
+        operator::{
+            binary::{ArithmeticOp, BinaryOp, BitwiseOp, LogicalOp, RelationalOp},
+            unary::UnaryOp,
+            Binary, Unary,
+        },
+    },
+    Expression,
+};
+use boa_interner::Interner;
+
+use super::{Pass, PassResult};
+
+#[derive(Debug, Default)]
+pub(crate) struct ConstantFoldingOptimizer {}
+
+impl ConstantFoldingOptimizer {
+    pub(crate) fn constant_fold_unary_expr(
+        unary: &mut Unary,
+        interner: &mut Interner,
+    ) -> PassResult<Expression> {
+        let value = if let Expression::Literal(Literal::Int(integer)) = unary.target() {
+            *integer
+        } else {
+            return PassResult::Leave;
+        };
+        let literal = match unary.op() {
+            UnaryOp::Minus => Literal::Int(-value),
+            UnaryOp::Plus => Literal::Int(value),
+            UnaryOp::Not => Literal::Bool(value == 0),
+            UnaryOp::Tilde => Literal::Int(!value),
+            UnaryOp::TypeOf => Literal::String(interner.get_or_intern("number")),
+            UnaryOp::Delete => Literal::Bool(true),
+            UnaryOp::Void => Literal::Undefined,
```

Comment:
Currently I'm only constant folding integer literals, and as mentioned in the previous reviews, `Literal::Undefined` is the an actual `undefined` value, not `undefined` global

Reply:
I may not be getting something, but wouldn't that cause incorrect foldings such as:
```javascript
{
  let a = 5;
  let undefined = 10;
  console.log(void a); // prints 10, should print 5
}
```

Reply:
Also, if you need to refer to the `undefined` value, you can add it to the list of static `Sym`s and refer to it with `Expression::Identifier(Sym::UNDEFINED.into())`.

Reply:
That is `Constant propagation` it is a another optimization, the are often grouped together but there different, `constant folding` optimization only deals with constant expressions.

Reply:
> Also, if you need to refer to the `undefined` value, you can add it to the list of static `Sym`s and refer to it with `Expression::Identifier(Sym::UNDEFINED.into())`.

But this gets the value from the global or local that can change, and `void 0` will always give undefined, irrelevant of other state, right?

Reply:
> But this gets the value from the global or local that can change, and `void 0` will always give undefined, irrelevant of other state, right?

Yep. I was just mentioning another way to get the `undefined` binding without adding an `Undefined` literal.

Reply:
> Yes, that is true, I was going to try to implement this after this optimization, since you simplify/fold the expression then you inline/propagate the constants

I mean that this current optimization will convert things such as:
```javascript
{
  let undefined = 10;
  console.log(void 5); // undefined
}
```
to
```javascript
{
  let undefined = 10;
  console.log(undefined); // 10
}
```
since `void 5` is in an expression node and constant folding will substitute it with the `undefined` binding, right?

I'm arguing that this is incorrect because it should instead evaluate to `JsValue::Undefined`, which can only be ensured if we used `void 0` directly:
```javascript
{
  let undefined = 10;
  console.log(void 0); // undefined
}
```

But now that I'm thinking it more thoroughly, wouldn't it be fine to leave the `void <integer>` as it is? Using `0` or `37373` as the rhs of the `void` operation shouldn't matter anyways.

Reply:
> since `void 5` is an expression node and constant folding will substitute it with the `undefined` binding, right?

It actually substitutes to `Literal::Undefined` witch is [mapped to `PushUndefined` opcode](https://github.com/boa-dev/boa/pull/2679/files#diff-620dfaf5e8f600ab363a3fa0d28280ed8946f91d36f86ed338f05c54cb2b19c2R36) and it pushes `JsValue::undefined()` directly, no binding lookup.

> But now that I'm thinking it more thoroughly, wouldn't it be fine to leave the `void <integer>` as it is? Using `0` or `37373` as the operand of the `void` operation shouldn't matter anyways.

It could be left like that but it would make folding more difficult also generate unneeded instructions.

Reply:
> It actually substitutes to `Literal::Undefined` witch is [mapped to `PushUndefined` opcode](https://github.com/boa-dev/boa/pull/2679/files#diff-620dfaf5e8f600ab363a3fa0d28280ed8946f91d36f86ed338f05c54cb2b19c2R36) and it pushes `JsValue::undefined()` directly, no binding lookup.

That's the detail I was missing! Thanks for explaining it :)